core/client/hdfs/src/main/java/alluxio/hadoop/AlluxioHdfsInputStream.java

```diff
@@ -50,7 +51,29 @@ public int read() throws IOException {
 
   @Override
   public int read(ByteBuffer buf) throws IOException {
-    return mInput.read(buf);
+    // @see <a href="https://github.com/apache/hadoop/blob/rel/release-3.3.6/
+    // * hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/
+    // * fs/FSDataInputStream.java#L154">FSDataInputStream.java</a>
+    if (mInput.getWrappedStream() instanceof ByteBufferReadable) {
+      return mInput.read(buf);
+    } else {
+      int off = buf.position();
+      int len = buf.remaining();
+      byte[] byteArray;
+      if (buf.hasArray()) {
+        byteArray = buf.array();
+      } else {
+        byteArray = new byte[len];
+      }
+
+      int totalBytesRead = read(byteArray);
+      if (totalBytesRead <= 0) {
+        return totalBytesRead;
+      }
+      buf.position(off).limit(off + len);
+      buf.put(byteArray, 0, totalBytesRead);
+      return totalBytesRead;
+    }
```

Comment:
You need to use `buf.arrayOffset()` to calculate the start offset of the array.

```suggestion
      int off = buf.position();
      int len = buf.remaining();
      final int totalBytesRead;
      if (buf.hasArray()) {
        byte[] byteArray = buf.array();
        totalBytesRead = read(byteArray, buf.arrayOffset() + off, len);
        if (totalBytesRead > 0) {
          buf.position(off + totalBytesRead);
        }
      } else {
        byte[] byteArray = new byte[len];
        totalBytesRead = read(byteArray);
        if (totalBytesRead > 0) {
          buf.put(byteArray, 0, totalBytesRead);
        }
      }
      return totalBytesRead;
    }
```